### PR TITLE
#1669 Rejection sampling follow-up

### LINKED ISSUE #1669: Rejection sampling follow-up
cf. https://github.com/JuliaReach/LazySets.jl/pull/1612

### PR DESCRIPTION
**Pull request by ueliwechsler**
Closes #1669.

### CODE REVIEW

**Comment by mforets** [comment]
The docs build fail because of

```julia
|┌ Warning: undefined binding 'LazySets._canonical_length' in `@docs` block in src/lib/utils.md:72-79
```
This function needs to be removed from the corresping file under `/docs` as well.

**Comment by mforets** [comment]
The proposed changes LGTM.

File: src/Utils/samples.jl [hunk 1]
```diff
@@ -120,63 +120,42 @@ end
...
-and redraws as long as the sample is not contained in the original set `X`,
-i.e., `x ∉ X`.
+Sampling of an arbitrary `LazySet` `X`.
 
 ### Input
 
 - `X`           -- lazyset
-- `num_samples`    -- number of random samples
+- `num_samples` -- number of random samples
+- `sampler`     -- Sampler used (default:`RejectionSampler`)
```
Comment (schillic) [inline src/Utils/samples.jl]:
```suggestion
- `sampler`     -- Sampler used (default: `RejectionSampler`)
```

File: src/Utils/samples.jl [hunk 2]
```diff
@@ -120,63 +120,42 @@ end
...
 - `rng`         -- (optional, default: `GLOBAL_RNG`) random number generator
 - `seed`        -- (optional, default: `nothing`) seed for reseeding
 
 ### Output
 
 A vector of `num_samples` vectors.
 If `num_samples` is not passed one sample as a single vector is returned.
+
+### Algorithm
+`RejectionSampler`:
```
Comment (schillic) [inline src/Utils/samples.jl]:
```suggestion

`RejectionSampler`:
```

File: src/Utils/samples.jl [hunk 3]
```diff
@@ -120,63 +120,42 @@ end
...
+Draw a sample `x` from a uniform distribution of a box-overapproximation of the
+original set `X` in all `n` dimensions. The function rejects a drawn sample `x`
+and redraws as long as the sample is not contained in the original set `X`,
+i.e., `x ∉ X`.
```
Comment (schillic) [inline src/Utils/samples.jl]: I suggest to use double ticks for math mode here (as in \`\`x\`\`) and to move this part to the documentation of `RejectionSampler` instead.

File: src/Utils/samples.jl [hunk 4]
```diff
@@ -185,35 +164,47 @@ function sample(X::LazySet{N}; kwargs...) where {N<:Real}
...
-    Sampler{S<:LazySet, D<:Distribution}
+    Sampler
+
+Abstract type for defining new sample methods.
+
+### Notes
+
+All subtypes should implement a `_sample_from_sampler!` method.
+"""
+abstract type Sampler end
```
[the remark this answers is not in the document]
Comment (schillic) [reply src/Utils/samples.jl]: I think `Sampler` is fine.

File: src/Utils/samples.jl [hunk 5]
```diff
@@ -185,35 +164,47 @@ function sample(X::LazySet{N}; kwargs...) where {N<:Real}
...
-Type used for Rejections Sampling from an arbitrary `LazySet`.
+Type used for rejection sampling of an arbitrary `LazySet` `X`.
 
 ### Fields
 
 - `X`           -- lazyset
 - `box_approx`  -- Distribution from which the sample is drawn
 """
-struct Sampler{S<:LazySet, D<:Distribution}
+struct RejectionSampler{S<:LazySet, D<:Distribution} <:Sampler
```
Comment (schillic) [inline src/Utils/samples.jl]:
```suggestion
struct RejectionSampler{S<:LazySet, D<:Distribution} <: Sampler
```

File: src/Utils/samples.jl [hunk 6]
```diff
@@ -185,35 +164,47 @@ function sample(X::LazySet{N}; kwargs...) where {N<:Real}
...
     box_approx::Vector{D}
 end
 
-function Sampler(X, distribution=Uniform)
-    box = _canonical_length(X)
-    box_approx = [distribution(box[i,:]...) for i = 1:size(box,1)]
-    return Sampler(X, box_approx)
+function RejectionSampler(X, distribution=Uniform)
+    B = box_approximation(X)
+    canonical_support =  hcat(low(B), high(B))
```
Comment (schillic) [inline src/Utils/samples.jl]:
```suggestion
    canonical_support = hcat(low(B), high(B))
```

File: src/Utils/samples.jl [hunk 7]
```diff
@@ -185,35 +164,47 @@ function sample(X::LazySet{N}; kwargs...) where {N<:Real}
...
+    dims = size(canonical_support,1)
```
Comment (schillic) [inline src/Utils/samples.jl]:
```suggestion
    dims = size(canonical_support, 1)
```

File: src/Utils/samples.jl [hunk 8]
```diff
@@ -185,35 +164,47 @@ function sample(X::LazySet{N}; kwargs...) where {N<:Real}
...
-    _rejection_sampling!(D::Vector{Vector{N}},
-           sampler::Sampler;
+    _sample_from_sampler!(D::Vector{Vector{N}},
+           sampler::RejectionSampler;
            rng::AbstractRNG=GLOBAL_RNG,
            seed::Union{Int, Nothing}=nothing) where {N<:Real}
 
-Rejection sampling from a box approximation of a set and distribution
-defined in `sampler`.
+Sampling points from a `Sampler`.
```
Comment (schillic) [inline src/Utils/samples.jl]: I think the documentation should say what this specific method does.
```suggestion
Sample points using rejection sampling.
```

File: src/Utils/samples.jl [hunk 9]
```diff
@@ -226,8 +217,8 @@ defined in `sampler`.
 
 A vector of `num_samples` vectors.
 """
-function _rejection_sampling!(D::Vector{Vector{N}},
-               sampler::Sampler;
+function _sample_from_sampler!(D::Vector{Vector{N}},
```
[the remark this answers is not in the document]
Comment (schillic) [reply src/Utils/samples.jl]: I support this.

File: docs/src/lib/utils.md [hunk 10]
```diff
@@ -72,8 +72,8 @@ implementing_sets
 ```@docs
 _sample_unit_nsphere_muller!
 _sample_unit_nball_muller!
-LazySets._canonical_length(X::LazySet{N}) where {N<:Real}
 sample
-LazySets.Sampler{S<:LazySet, D<:Distribution}
-LazySets._rejection_sampling!
+LazySets.Sampler
+LazySets.RejectionSampler{S<:LazySet, D<:Distribution}
+LazySets._sample_from_sampler!
```
Comment (mforets) [inline docs/src/lib/utils.md]:
```suggestion
LazySets._sample!
```

Comment (mforets) [inline docs/src/lib/utils.md]:
```suggestion
LazySets._sample!
```

**Comment by mforets** [comment]
Thanks a lot @ueliwechsler !

**Pull request merged**